### PR TITLE
check if element exists before creation

### PR DESCRIPTION
**Pull request by AbdelHedhili**
Signed-off-by: Abdelsalem <abdelsalem.hedhili@rte-france.com>

### CODE REVIEW

File: src/main/java/org/gridsuite/directory/server/DirectoryService.java [hunk 1]
```diff
@@ -117,12 +117,17 @@ private void emitDirectoryChanged(UUID directoryUuid, String userId, String erro
...
-        return insertElement(elementAttributes, parentDirectoryUuid).doOnSuccess(unused -> emitDirectoryChanged(
-            parentDirectoryUuid,
-            userId,
-            isPrivateForNotification(parentDirectoryUuid, elementAttributes.getAccessRights().isPrivate()),
-            false,
-            NotificationType.UPDATE_DIRECTORY));
+        if (elementExistsInDirectory(parentDirectoryUuid, elementAttributes.getElementName(), elementAttributes.getType())) {
+            return Mono.error(new DirectoryException(NOT_ALLOWED));
+        } else {
+            return insertElement(elementAttributes, parentDirectoryUuid).doOnSuccess(unused2 -> emitDirectoryChanged(
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/directory/server/DirectoryService.java:123]: unused2 ? why 2 ?

File: src/main/java/org/gridsuite/directory/server/DirectoryService.java [hunk 2]
```diff
@@ -265,9 +270,12 @@ private boolean isPrivateForNotification(UUID parentDirectoryUuid, boolean isCur
         }
     }
 
-    public Mono<Void> elementExists(UUID parentDirectoryUuid, String elementName, String userId) {
-        return directoryElementRepository.findDirectoryContentByUserId(parentDirectoryUuid, userId).stream().anyMatch(e -> e.getName().equals(elementName))
-            ? Mono.empty() : Mono.error(new DirectoryException(NOT_FOUND));
+    public boolean elementExistsInDirectory(UUID parentDirectoryUuid, String elementName, String type) {
+        return (long) directoryElementRepository.findByNameAndParentIdAndType(elementName, parentDirectoryUuid, type).size() > 0;
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/directory/server/DirectoryService.java:274]: you change parameters order here, no big deal

File: src/main/java/org/gridsuite/directory/server/DirectoryController.java [hunk 3]
```diff
@@ -111,14 +112,14 @@ public DirectoryController(DirectoryService service) {
         return ResponseEntity.ok().body(service.deleteElement(elementUuid, userId));
     }
 
-    @RequestMapping(method = RequestMethod.HEAD, value = "/directories/{directoryUuid}/{elementName}")
-    @Operation(summary = "Check if the element exists")
+    @RequestMapping(method = RequestMethod.HEAD, value = "/directories/{directoryUuid}/{elementName}/{type}")
+    @Operation(summary = "Check if the element exists in the given directory")
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/directory/server/DirectoryController.java:116]: `Check if an element with this name and this type already exists in the given directory `?

File: src/main/java/org/gridsuite/directory/server/DirectoryController.java [hunk 4]
```diff
@@ -111,14 +112,14 @@ public DirectoryController(DirectoryService service) {
...
     @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The element exists"),
                            @ApiResponse(responseCode = "404", description = "The element doesn't exist")})
     public ResponseEntity<Mono<Void>> elementExists(@PathVariable("directoryUuid") UUID directoryUuid,
                                                     @PathVariable("elementName") String elementName,
-                                                    @RequestHeader("userId") String userId) {
-        return ResponseEntity.ok().body(service.elementExists(directoryUuid, elementName, userId));
+                                                    @PathVariable("type") String type) {
+        return ResponseEntity.ok().body(service.elementExistsMono(directoryUuid, elementName, type));
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/directory/server/DirectoryController.java:122]: HEAD request method shouldn't return a body ! It was like that before but it's not good.
cf. warning : https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD

**Review by sBouzols — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/directory/server/DirectoryService.java (hunk 2 above)
Comment (etiennehomer) [inline src/main/java/org/gridsuite/directory/server/DirectoryService.java:274]: Why this cast to (long) ?

**Review by sBouzols — APPROVED** [review]

File: src/main/java/org/gridsuite/directory/server/DirectoryService.java [hunk 5]
```diff
@@ -112,12 +112,17 @@ private void emitDirectoryChanged(UUID directoryUuid, String userId, String erro
...
     /* methods */
     public Mono<ElementAttributes> createElement(ElementAttributes elementAttributes, UUID parentDirectoryUuid, String userId) {
-        return insertElement(elementAttributes, parentDirectoryUuid).doOnSuccess(unused -> emitDirectoryChanged(
-            parentDirectoryUuid,
-            userId,
-            isPrivateForNotification(parentDirectoryUuid, elementAttributes.getAccessRights().isPrivate()),
-            false,
-            NotificationType.UPDATE_DIRECTORY));
+        if (elementExistsInDirectory(parentDirectoryUuid, elementAttributes.getElementName(), elementAttributes.getType())) {
+            return Mono.error(new DirectoryException(NOT_ALLOWED));
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/directory/server/DirectoryService.java:116]: Add constant ALREADY_EXISTS ?

**Review by etiennehomer — APPROVED** [review]

**Pull request merged**